### ISSUE TITLE
clang-tidy crash for clang-analyzer-fuchsia.HandleChecker

### ISSUE THREAD

**Issue by ottojo** [llvm/llvm-project#57011]
Commandline:
```shell
clang-tidy src/handle_attributes.cpp -checks=-\*,clang-analyzer-fuchsia.HandleChecker -- --std=c++20
```

Input source:
```cpp
struct zx_handle_t{};

[[clang::acquire_handle("Fuchsia")]] zx_handle_t open_thing();

void close_thing(zx_handle_t handle [[clang::release_handle("Fuchsia")]]) noexcept;

void use_thing(zx_handle_t handle [[clang::use_handle("Fuchsia")]]);

void doStuff() {
    zx_handle_t handle = open_thing();
    close_thing(handle);
    close_thing(handle);
}
```

Backtrace:
```console
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: clang-tidy src/handle_attributes.cpp -checks=-*,clang-analyzer-fuchsia.HandleChecker -- --std=c++20
1.      <eof> parser at end of file
2.      While analyzing stack: 
        #0 Calling doStuff()
3.      /home/jonas/Workspace/sandbox/src/handle_attributes.cpp:10:5: Error evaluating statement
 #0 0x00007f21a32d545a (/usr/lib/libLLVM-14.so+0xcd545a)
 #1 0x00007f21a32d2aef llvm::sys::RunSignalHandlers() (/usr/lib/libLLVM-14.so+0xcd2aef)
 #2 0x00007f21a32d2c3d (/usr/lib/libLLVM-14.so+0xcd2c3d)
 #3 0x00007f21a2054a40 (/usr/lib/libc.so.6+0x38a40)
 #4 0x00007f21ab8817e0 clang::ento::SymbolReaper::isLive(clang::ento::SymExpr const*) (/usr/lib/libclang-cpp.so.14+0x20817e0)
 #5 0x00007f21ab946b47 (/usr/lib/libclang-cpp.so.14+0x2146b47)
 #6 0x00007f21ab75e6b6 clang::ento::CheckerManager::runCheckersForDeadSymbols(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNodeSet const&, clang::ento::SymbolReaper&, clang::Stmt const*, clang::ento::ExprEngine&, clang::ProgramPoint::Kind) (/usr/lib/libclang-cpp.so.14+0x1f5e6b6)
 #7 0x00007f21ab7c6919 clang::ento::ExprEngine::removeDead(clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&, clang::Stmt const*, clang::LocationContext const*, clang::Stmt const*, clang::ProgramPoint::Kind) (/usr/lib/libclang-cpp.so.14+0x1fc6919)
 #8 0x00007f21ab7daabc clang::ento::ExprEngine::ProcessStmt(clang::Stmt const*, clang::ento::ExplodedNode*) (/usr/lib/libclang-cpp.so.14+0x1fdaabc)
 #9 0x00007f21ab7ddbaf clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) (/usr/lib/libclang-cpp.so.14+0x1fddbaf)
#10 0x00007f21ab7890ff clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) (/usr/lib/libclang-cpp.so.14+0x1f890ff)
#11 0x00007f21ab789308 clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) (/usr/lib/libclang-cpp.so.14+0x1f89308)
#12 0x00007f21abb5ea20 (/usr/lib/libclang-cpp.so.14+0x235ea20)
#13 0x00007f21abb6764d (/usr/lib/libclang-cpp.so.14+0x236764d)
#14 0x00007f21ab5bfe50 clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) (/usr/lib/libclang-cpp.so.14+0x1dbfe50)
#15 0x00007f21a9fd9efb clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/libclang-cpp.so.14+0x7d9efb)
#16 0x00007f21ab5b39b9 clang::FrontendAction::Execute() (/usr/lib/libclang-cpp.so.14+0x1db39b9)
#17 0x00007f21ab552e67 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/libclang-cpp.so.14+0x1d52e67)
#18 0x00007f21ab701528 clang::tooling::FrontendActionFactory::runInvocation(std::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) (/usr/lib/libclang-cpp.so.14+0x1f01528)
#19 0x00005632543db198 (/usr/bin/clang-tidy+0x48d198)
#20 0x00007f21ab6e8f87 clang::tooling::ToolInvocation::runInvocation(char const*, clang::driver::Compilation*, std::shared_ptr<clang::CompilerInvocation>, std::shared_ptr<clang::PCHContainerOperations>) (/usr/lib/libclang-cpp.so.14+0x1ee8f87)
#21 0x00007f21ab7007d4 clang::tooling::ToolInvocation::run() (/usr/lib/libclang-cpp.so.14+0x1f007d4)
#22 0x00007f21ab702220 clang::tooling::ClangTool::run(clang::tooling::ToolAction*) (/usr/lib/libclang-cpp.so.14+0x1f02220)
#23 0x00005632543df17a clang::tidy::runClangTidy(clang::tidy::ClangTidyContext&, clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>, bool, bool, llvm::StringRef) (/usr/bin/clang-tidy+0x49117a)
#24 0x0000563254053c11 clang::tidy::clangTidyMain(int, char const**) (/usr/bin/clang-tidy+0x105c11)
#25 0x00007f21a203f2d0 (/usr/lib/libc.so.6+0x232d0)
#26 0x00007f21a203f38a __libc_start_main (/usr/lib/libc.so.6+0x2338a)
#27 0x000056325404a915 _start (/usr/bin/clang-tidy+0xfc915)
[1]    187653 segmentation fault (core dumped)  clang-tidy src/handle_attributes.cpp  -- --std=c++20
```

Version:
```console
$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 14.0.6
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: znver2
```

**Comment by haoNoQ** [comment]
@Xazax-hun this one's yours!

The backtrace is somewhat surprising, are we putting a null symbol into `SymbolReaper` or is the `SymbolReaper` itself broken?

**Comment by ottojo** [comment]
(Some context: This only appears if i name the handle `zx_handle_t`, which is hardcoded here https://github.com/llvm/llvm-project/blob/main/clang/lib/StaticAnalyzer/Checkers/FuchsiaHandleChecker.cpp#L111. I found the handle attributes and decided to poke at it, not realizing that this probably only works for fuchsia stuff...)

**Comment by haoNoQ** [comment]
Aha ok, I do not recommend mimicking Fuchsia interfaces in non-Fuchsia code just to take advantage of the checker. The struct name isn't the only thing that's hardcoded in the checker, the semantics behind the interface also is. Even relatively minor differences from the original interfaces can lead to very unexpected results. If you're not writing Fuchsia code and these interfaces don't come from official Fuchsia headers, you shouldn't use these checkers. It should be possible to _adapt_ these checkers to your use case but out of the box they're probably not suitable for that.

We still shouldn't crash though.

**Comment by ottojo** [comment]
Absolutely. It might be beneficial to mention in https://clang.llvm.org/docs/AttributeReference.html#handle-attributes that those attributes are currently only used by a Fuchsia-specific analyzer.

**Comment by Xazax-hun** [comment]
Whooopsie. Sorry, it looks like I GitHub decided to not to send me notifications from this repo. I'll try to take a look at this during the weekend.

**Comment by Xazax-hun** [comment]
I tried to reproduce it and the released version (14.0.6) reproduced the crash, but I was unable to reproduce it using a fresh checkout of main as of yesterday. I suppose that this particular crash has been fixed independently. Feel free to reopen if you can still reproduce it with main or the next version.
